banderwagon/element.go

```diff
@@ -40,16 +41,21 @@ type Element struct {
 
 // Bytes returns the compressed serialized version of the element.
 func (p Element) Bytes() [CompressedSize]byte {
-	// Convert underlying point to affine representation.
-	var affine bandersnatch.PointAffine
-	affine.FromProj(&p.inner)
-
 	// Serialisation takes the x co-ordinate and multiplies it by the sign of y.
-	x := affine.X
-	if !affine.Y.LexicographicallyLargest() {
-		x.Neg(&x)
+	affineX := p.inner.X
+	affineY := p.inner.Y
+	if !p.inner.Z.IsOne() {
+		// Convert underlying point to affine representation.
+		var affine bandersnatch.PointAffine
+		affine.FromProj(&p.inner)
+		affineX = affine.X
+		affineY = affine.Y
 	}
-	return x.Bytes()
+
+	if !affineY.LexicographicallyLargest() {
+		affineX.Neg(&affineX)
+	}
+	return affineX.Bytes()
```

Comment:
While doing the benchmarks I realized there was an unfortunate fact here.

In the previous version in this diff, we always did `affine.FromProj(&p.inner)` (L45).
Unfortunately, this means that we'll [always calculate the inverse](https://github.com/Consensys/gnark-crypto/blob/master/ecc/bls12-381/bandersnatch/point.go#L241) of `Z`. And if you see [the inverse implementation](https://github.com/Consensys/gnark-crypto/blob/master/ecc/bls12-381/fr/element.go#L1255), there's no "short-circuit" for the case of `Z=1`

This is relevant since in multiproof, we now normalize all `Cs` points and then we do these `Bytes()` calls to feed FS. So doing "inverse" calculations of `Z=1` is wasted work.

What I did here is nothing more than doing the L47 check, and only calling `affine.FromProj` if needed. This won't be the case when generating proofs, but probably in other calls is clearly relevant. So only do the work if necessary.